src/adapter/src/coord.rs

```diff
@@ -2131,178 +2140,217 @@ impl Coordinator {
+        // frontiers, so if we can reconstruct them, we know that reconciliation will succeed. Note
+        // that to reconstruct a warmup frontier, we must only take into account the frontiers of
+        // (transitive) storage dependencies: The frontiers of index dependencies can regress
+        // across restarts, so we'd risk ending up with a warmup frontier that is too early to
+        // ensure successful reconciliation.
+        //
+        // Note that choosing the warmup frontier only based on storage dependencies means that the
+        // dataflow might have to wait for indexes in between to catch up to this frontier, if
+        // their `as_of` is selected earlier.
+        let storage_dependencies = storage_constraints.dependencies_bundle();
+        let warmup_frontier = self.greatest_available_read(&storage_dependencies);
+
+        // Apply additional constraints based on whether this dataflow exports an index or a
+        // materialized view.
+        let write_frontier;
+        let candidate_as_of = if dataflow.exported_index_ids().next().is_some() {
+            // Index dataflow.
+
+            write_frontier = self.least_valid_write(&storage_dependencies);
+            if let Some(ts) = write_frontier.as_option() {
+                // If the index has a compaction window configured, we should hold back the `as_of`
+                // to ensure this window is queryable after the index was created. Doing so should
+                // not break compute reconciliation because the compute controller should have
+                // prevented replicas from compacting their installed dataflows into the compaction
+                // window.
+                let max_compaction_frontier = Antichain::from_elem(compaction_window.lag_from(*ts));
+                soft_assert_or_log!(
+                    !max_compaction_frontier.is_empty(),
+                    "`max_compaction_frontier` unexpectedly empty",
+                );
+                max_compaction_frontier
+            } else {
+                // The write frontier is empty. This can happen for constant collections, and for
+                // an index on a REFRESH MV that is past its last refresh. Installing an index with
+                // an empty frontier is not useful since that index would not be readable, so we
+                // bail using the minimum frontier instead. This is something we could refine in
+                // the future.
+                min_as_of.clone()
+            }
+        } else if let Some(sink_id) = dataflow.persist_sink_ids().next() {
+            // Materialized view dataflow.
+
+            write_frontier = self.storage_write_frontier(sink_id).clone();
+            // Materialized view dataflows are only depended on by their target storage collection,
+            // so `write_frontier` and `max_as_of` should always be equal.
+            soft_assert_eq_or_log!(write_frontier, max_as_of);
```

Comment:
This is not exactly true: when the MV's storage write frontier is still at 0, the `max_as_of` will be >0, because of the `join` in the `max_as_of` determination.

Maybe just assert that `PartialOrder::less_equal(&write_frontier, &max_as_of)`. This would still cover my original suggestion (if `write_frontier.is_empty()` then also `max_as_of.is_empty()`).

Reply:
(Btw. `3d.` in `workflow_test_refresh_mv_restart` catches this, but it's not run in CI at the moment, because it's ignored due to an unrelated bug making some other test interfere with it. But it's possible to run it just by itself locally: `bin/mzcompose --dev --find cluster run test-refresh-mv-restart`)

Reply:
Uff, good catch. Fixed!